### PR TITLE
QGIS Project 3 Edition config compatibility

### PR DESCRIPTION
**Pull request by rldhont**
Compatibility with QGIS3 #1075

### CODE REVIEW

**Comment by rldhont** [comment]
@rmarzocchi84 I have cloned your message, to fix the code here.

From a preliminary test it seems to work correctly. 

First of all I test:
- Value Map
- Datetime field


I only find a strange behaviour that perhaps depends on a QGIS bug:


This value map seems not correctly defined 
    

` <Option name="map" type="Map">
`

 and do not works:



    <field name="d_origine">
          <editWidget type="ValueMap">
            <config>
              <Option type="Map">
                <Option name="map" type="Map">
                  <Option name="Altro" value="5" type="QString"/>
                  <Option name="Atti amministrativi" value="3" type="QString"/>
                  <Option name="Palmare" value="4" type="QString"/>
                  <Option name="Rilievo manuale" value="1" type="QString"/>
                  <Option name="Rilievo meccanico" value="2" type="QString"/>
                </Option>
              </Option>
            </config>
          </editWidget>
        </field>
		
	

This value map is correctly defined 

` <Option name="map" type="List">
`

and works:	
		
		
		
		<field name="d_origine">
          <editWidget type="ValueMap">
            <config>
              <Option type="Map">
                <Option name="map" type="List">
                  <Option type="Map">
                    <Option name="Altro" value="5" type="QString"/>
                  </Option>
                  <Option type="Map">
                    <Option name="Atti amministrativi" value="3" type="QString"/>
                  </Option>
                  <Option type="Map">
                    <Option name="Palmare" value="4" type="QString"/>
                  </Option>
                  <Option type="Map">
                    <Option name="Rilievo manuale" value="1" type="QString"/>
                  </Option>
                  <Option type="Map">
                    <Option name="Rilievo meccanico" value="2" type="QString"/>
                  </Option>
                </Option>
              </Option>
            </config>
          </editWidget>
        </field>


After I test *Value Relation* and it do not works. This is the QGIS project definition:

        <field name="cod_strada">
          <editWidget type="ValueRelation">
            <config>
              <Option type="Map">
                <Option name="AllowMulti" value="false" type="bool"/>
                <Option name="AllowNull" value="false" type="bool"/>
                <Option name="FilterExpression" value="" type="QString"/>
                <Option name="Key" value="cod_strada" type="QString"/>
                <Option name="Layer" value="route20190320155525256" type="QString"/>
                <Option name="NofColumns" value="1" type="int"/>
                <Option name="OrderByValue" value="true" type="bool"/>
                <Option name="UseCompleter" value="true" type="bool"/>
                <Option name="Value" value="cod_strada" type="QString"/>
              </Option>
            </config>
          </editWidget>
        </field>

**Comment by rldhont** [comment]
@rmarzocchi84 The commit https://github.com/3liz/lizmap-web-client/pull/1202/commits/0c3e7b151d90e5e05de2ed3be114fef896fa7db0 fixes this config:

    <field name="d_origine">
      <editWidget type="ValueMap">
        <config>
          <Option type="Map">
            <Option name="map" type="Map">
              <Option name="Altro" value="5" type="QString"/>
              <Option name="Atti amministrativi" value="3" type="QString"/>
              <Option name="Palmare" value="4" type="QString"/>
              <Option name="Rilievo manuale" value="1" type="QString"/>
              <Option name="Rilievo meccanico" value="2" type="QString"/>
            </Option>
          </Option>
        </config>
      </editWidget>
    </field>

**Comment by rldhont** [comment]
@rmarzocchi84 I have tested ValueRelation, and it works in montpellier project. I will wait for more review.

**Comment by rmarzocchi84** [comment]
> @rmarzocchi84 I have tested ValueRelation, and it works in montpellier project. I will wait for more review.

I will check better ValueRelation, and also your latest fix!

**Comment by rmarzocchi84** [comment]
The ValueRelation works, but the combobox do not show the small arrow button on the right. It is normal?

image

File: lizmap/modules/lizmap/classes/qgisFormControl.class.php [hunk 1]
```diff
@@ -258,7 +261,11 @@ public function __construct($ref, $edittype, $prop, $aliasXml = null, $defaultVa
...
             } elseif ($this->fieldEditType === 'TextEdit') {
-                if (property_exists($this->widgetv2configAttr, 'UseHtml')) {
+                if (property_exists($this->widgetv2configAttr, 'IsMultiline')
+                    && ($this->widgetv2configAttr->IsMultiline === '0' || $this->widgetv2configAttr->IsMultiline === 'false')) {
+                    $this->fieldEditType = 'LineEdit';
+                    $markup = $this->qgisEdittypeMap[$this->fieldEditType]['jform']['markup'];
+                } else if (property_exists($this->widgetv2configAttr, 'UseHtml')) {
                     $markup = $this->qgisEdittypeMap[$this->fieldEditType]['jform']['markup'][(int) $this->widgetv2configAttr->UseHtml];
                 } else {
                     $markup = $this->qgisEdittypeMap[$this->fieldEditType]['jform']['markup'][0];
```
Comment (nboisteault) [inline lizmap/modules/lizmap/classes/qgisFormControl.class.php]: @rldhont line 276 : allowMulti is now a boolean. Changing cast (int) to (bool) works.

```xml
        <field name="name">
          <editWidget type="ValueRelation">
            <config>
              <Option type="Map">
                <Option type="bool" name="AllowMulti" value="true"/>
                <Option type="bool" name="AllowNull" value="true"/>
                <Option type="QString" name="FilterExpression" value=""/>
                <Option type="QString" name="Key" value="id"/>
                <Option type="QString" name="Layer" value="test_layer_7714a8b5_23f3_4fe1_a100_af6b23df6940"/>
                <Option type="int" name="NofColumns" value="1"/>
                <Option type="bool" name="OrderByValue" value="false"/>
                <Option type="bool" name="UseCompleter" value="false"/>
                <Option type="QString" name="Value" value="name"/>
              </Option>
            </config>
          </editWidget>
        </field>
```

**Comment by rldhont** [comment]
> The ValueRelation works, but the combobox do not show the small arrow button on the right. It is normal?
> 
> image

Yes, it is normal.

File: lizmap/modules/lizmap/classes/qgisFormControl.class.php (hunk 1 above)
Comment (rldhont) [reply lizmap/modules/lizmap/classes/qgisFormControl.class.php]: AllowMulti is not yet used

Comment (nboisteault) [reply lizmap/modules/lizmap/classes/qgisFormControl.class.php]: @rldhont Allowmulti was supported.
Here is my fix, it works with '0', '1', 'false' and 'true' : 
`$markup = $this->qgisEdittypeMap[$this->fieldEditType]['jform']['markup'][(int) filter_var($this->widgetv2configAttr->AllowMulti, FILTER_VALIDATE_BOOLEAN) ];`

Comment (rldhont) [reply lizmap/modules/lizmap/classes/qgisFormControl.class.php]: Done

**Comment by rldhont** [comment]
@nboisteault I have force the branch to integrate color widget.
Reload the branch from master

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 2]
```diff
@@ -904,16 +993,12 @@ private function fillControlFromValueRelationLayer($fieldName, $formControl)
             $dataSource = new jFormsStaticDatasource();
 
             // required
-            if (
-                strtolower($formControl->valueRelationData['allowNull']) == 'false'
-                or
-                strtolower($formControl->valueRelationData['allowNull']) == '0'
-            ) {
+            if (!$formControl->valueRelationData['allowNull']) {
```
[the remark this answers is not in the document]
Comment (nboisteault) [reply lizmap/modules/lizmap/classes/qgisForm.class.php]: Patch coming.

**Comment by nboisteault** [comment]
@laurentj jquery write a warning in console for color picker :
`The specified value "" does not conform to the required format.  The format is "#rrggbb" where rr, gg, bb are two-digit hexadecimal numbers.`

`value` should not be empty as written [here](https://developer.mozilla.org/fr/docs/Web/HTML/Element/Input/color#Valeur) :

> La valeur respecte nécessairement ce format et n'est jamais vide.

**Comment by rldhont** [comment]
@nboisteault is it the same in QGIS ?

**Comment by nboisteault** [comment]
@rldhont there is a default color #ff0000 when you open the picker.
Capture du 2019-04-17 10-49-59

**Comment by rldhont** [comment]
@nboisteault color picker has been added with an other PR. You can open an issue dedicated to empty color field.

**Comment by rmarzocchi84** [comment]
I test it today an all seems to works fine, also file upload! This update is really appreciated! 

There is also this feature (https://github.com/3liz/lizmap-web-client/issues/1027) which can be useful for end users.

**Pull request merged**